Initial code refactoring in preparation for other analyses

blinkseb opened the pull request:
Hi everyone,

I'm the JERC contact for the development of the common framework. I've started to do some work on this framework, and this PR is the first step: a (large) code refactoring.

I've not added any functionality, but mainly reworked the code structure to ease the inclusion of other analyzers. The final aim is to have one class per object, dumping informations into a tree (one tree per object). For the moment, only PUPPI and Jets are present, but we'll also need photons, met, muons and electrons for the others studies (I'll contact each JERC analyses to know what are their special needs here).

In this PR you'll also find:
  * Switch from getByLabel to getByToken
  * No more hardcoded InputTag in the code (read everything from the python configuration)

Also note that I've used a library of mine (https://github.com/blinkseb/TreeWrapper) to create the trees, you'll find it very easy to use :) If you want to test the PR, you'll need to grab this library:

```sh
git clone git@github.com:blinkseb/TreeWrapper.git JMEAnalysis/TreeWrapper
scram b -j8
```
I've also removed anything related to the reading of produced root file. All the analyses already have their own code to do that so it's redundant and hard to maintain.

Please don't hesitate if you have any remarks or concerns!

Cheers,
Sébastien

aperloff commented on src/JetMETAnalyzer.cc:
```diff
@@ -0,0 +1,517 @@
+  edm::EDGetTokenT<std::vector<pat::Jet>> srcJet_;
+  edm::EDGetTokenT<double> srcRho_;
+  edm::EDGetTokenT<std::vector<reco::Vertex>> srcVtx_;
+  edm::EDGetTokenT<std::vector<pat::Muon>> srcMuons_;
```
Will this work with the two >> right next to each other? Shouldn't there be a space between them?

blinkseb replied:
It's a valid syntax since C++11

aperloff replied:
Thank you, good to know. I didn't realize ROOT/CMSSW was C++11 compliant.

rappoccio replied:
Only ROOT6 is. I would suggest keeping the spaces for backwards compatibility.

blinkseb replied:
Yes but CMSSW is built with C++11 support since 5.3.10+ so it should be safe, right? I don't think ROOT will ever parse this file.

aperloff commented on README.md:
````diff
@@ -1,4 +1,13 @@
+```sh
+git clone git@github.com:blinkseb/TreeWrapper.git JMEAnalysis/TreeWrapper
+scram b -j8
+```
````
Can you add the ability to set the branch status or get a specific entry to your TreeWrapper class? This will certainly become useful if/when these trees become large. Also, at least one of my JEC codes already requires the SetBranchStatus method. I know I'm looking to the future here.

blinkseb replied:
Yep sure.
For the SetBranchStatus, currently, all the branches are disabled by default (status to 0) and only enabled when a read() is requested.

aperloff commented on src/JetMETAnalyzer.cc:
```diff
@@ -0,0 +1,517 @@
+  std::vector<float>& nCh = tree["nCh"].write<std::vector<float>>();
+  std::vector<float>& nNeutrals = tree["nNeutrals"].write<std::vector<float>>();
+  std::vector<float>& ptD = tree["ptD"].write<std::vector<float>>();
+  std::vector<bool>& isMatched = tree["isMatched"].write<std::vector<bool>>();
```
Part of what I liked about the ntuple solution (previous style) was that it removed the need for long lists of variables to be added directly into the code. Instead they were kept elsewhere and access from the "ntuple->" method. Is there any way to do that here? i.e. Can we move these to a new file or otherwise merge them into a new object? If not then this is fine; I think overall this is a huge improvement.

blinkseb replied:
I also don't like the long list, but didn't found another way when writing the code. I now think I may have a solution here, but I'll need to think a bit for the implementation.

I'll update this PR when it's done.

aperloff commented on src/JetMETAnalyzer.cc:
```diff
@@ -0,0 +1,517 @@
+  nCh.clear();
+  nNeutrals.clear();
+  ptD.clear();
+  isMatched.clear();
```
If all of these clears happen each event it might be nice to move these to a beginEvent() function (or any other function that is not the analyze function). Although, that depends on either moving the references to the class declaration or a new object, as suggested above. Thoughts?

blinkseb replied:
I'm pretty sure I can automatize the 'reset' at the beginning of each event. I'd be nice to not have to worry of resetting the state of each variable and having the wrapper to do that automatically.

aperloff commented on src/JetMETAnalyzer.cc:
```diff
@@ -0,0 +1,517 @@
+
+         float deta = part->eta() - etaJet;
+         float dphi = 2*atan(tan((part->phi()-phiJet)/2));
+         float dR = sqrt(deta*deta + dphi*dphi);
```
There are standard packages to do the deltaR calculation ("DataFormats/Math/interface/deltaR.h" --> reco::deltaR). Should we be using those? Saves a line or two.

blinkseb replied:
Yep will do.

aperloff commented on src/JetMETAnalyzer.cc:
```diff
@@ -0,0 +1,517 @@
+                 }
+             }
+         }
+     }
```
Can this whole beta/betaStar loop be moved to a new function? It just gets it out of the main analyzer area and makes reading the code a little easier. I realize there might be some issues with passing references or getting the output as needed below.

blinkseb replied:
Yep sure.

I just used the previous code without doing any modifications, but I'll clean it up and update this PR.

Thanks for the detailed review :)

rappoccio commented:
Do we really need a home-cooked thin wrapper layer with only a slightly different interface than standard ROOT? If people are desperate, then fine, but I don't personally see the need. 

It also uses a string-based map which is going to be abysmally slow. At the very least, use some kind of integer indexing with a hash or we'll be spending the next several years primarily waiting for string comparisons ;).

blinkseb commented:
I personally find the ROOT interface for dealing with trees absolutely awful, so the wrapper ;)

For the string-based map, hashing is already done by the `unordered_map` (fortunately!), and from my benchmarks, I've not seen any performance penalty compared to all the stuff done by CMSSW :)

rappoccio commented:
Can you give the timing differences of the module before and after your change, and also compare the TTrees at the end bit-by-bit so we can be sure there are no bugs?

blinkseb commented:
No problem, I'll prepare that!

rappoccio commented:
Thanks!

blinkseb commented:
I've updated the code taking into account feedback from @aperloff and @rappoccio.

Main changes are:
- Branch declaration are moved into the header file, thanks to C++11 member initialization feature.
- All branches are automatically reset to a default state after filling the tree. Default means 0, expect for vector where it's a clear().

With theses changes, the overhead of the library is null, since everything is done when the class is created and the tree registered? There's no map access in the event loop.

Here are the timings information for the old code:

```
TimeReport   0.001710   0.001714   0.001710   0.001714 puppiReader
TimeReport   0.000490   0.000492   0.000490   0.000492 nt_AK4PFchs
TimeReport   0.000431   0.000434   0.000431   0.000434 nt_AK8PFchs
TimeReport   0.000284   0.000294   0.000284   0.000294 nt_AK4PUPPI
TimeReport   0.000252   0.000253   0.000252   0.000253 nt_AK8PUPPI
TimeReport   0.001223   0.001245   0.001223   0.001245 nt_AK4PF
TimeReport   0.001245   0.001281   0.001245   0.001281 nt_AK8PF
```

 and the new one :

```
TimeReport   0.001728   0.001732   0.001728   0.001732 puppiReader
TimeReport   0.000319   0.000321   0.000319   0.000321 jmfw_AK4PFchs
TimeReport   0.000327   0.000328   0.000327   0.000328 jmfw_AK8PFchs
TimeReport   0.000122   0.000123   0.000122   0.000123 jmfw_AK4PFPUPPI
TimeReport   0.000118   0.000120   0.000118   0.000120 jmfw_AK8PFPUPPI
TimeReport   0.000820   0.000821   0.000820   0.000821 jmfw_AK4PF
TimeReport   0.000923   0.000925   0.000923   0.000925 jmfw_AK8PF
```

The timing is identical for puppiReader, and we see a gain with the new code for the JetMETAnalyzer. To be honest, I think it's mostly due to the fact that I removed some `cout` printing stuff for every event ... 

For the comparison of the data, I plotted each branch for the new code vs the old one. You can find all the plots here: https://sbrochet.web.cern.ch/sbrochet/#files%2FJME%20Framework%2FValidation
All the variables are computed the same way as in the old code.

rappoccio commented:
Hi, Sebastien,

OK, looks good! Thanks for doing this so quickly. 

Cheers,
Sal
